src/experiments/rl_envs.jl

```diff
@@ -1178,3 +1179,247 @@ function RLCore.Experiment(
+        end,
+    )
+
+    description = Description("# Play CartPole with VPG", save_dir)
+
+    Experiment(agent, env, stop_condition, hook, description)
+end
+
+function RLCore.Experiment(
+    ::Val{:JuliaRL},
+    ::Val{:VPG},
+    ::Val{:PendulumD},
+    ::Nothing;
+    save_dir = nothing,
+    seed = 2213,
+)
+    if isnothing(save_dir)
+        t = Dates.format(now(), "yyyy_mm_dd_HH_MM_SS")
+        save_dir = joinpath(pwd(), "checkpoints", "JuliaRL_VPG_PendulumD_$(t)")
+    end
+
+    lg = TBLogger(joinpath(save_dir, "tb_log"), min_level = Logging.Info)
+    rng = MersenneTwister(seed)
+    env = PendulumEnv(;
+        T = Float32,
+        rng = rng,
+        continuous = false,
+        n_actions = 3,
+        max_steps = 100,
+    )
+    ns, na = length(get_state(env)), length(get_actions(env))
+
+    agent = Agent(
+        policy = VPGPolicy(
+            approximator = NeuralNetworkApproximator(
+                model = Chain(
+                    Dense(ns, 128, relu; initW = glorot_uniform(rng)),
+                    Dense(128, 128, relu; initW = glorot_uniform(rng)),
+                    Dense(128, na; initW = glorot_uniform(rng)),
+                ),
+                optimizer = ADAM(),
+            ) |> cpu,
+            baseline = NeuralNetworkApproximator(
+                model = Chain(
+                    Dense(ns, 128, relu; initW = glorot_uniform(rng)),
+                    Dense(128, 128, relu; initW = glorot_uniform(rng)),
+                    Dense(128, 1; initW = glorot_uniform(rng)),
+                ),
+                optimizer = ADAM(),
+            ) |> cpu,
+            γ = 0.99f0,
+            rng = rng,
+        ),
+        trajectory = ElasticCompactSARTSATrajectory(
+            state_type = Float32,
+            state_size = (ns,),
+        ),
+    )
+
+    total_reward_per_episode = TotalRewardPerEpisode()
+    time_per_step = TimePerStep()
+    hook = ComposedHook(
+        total_reward_per_episode,
+        time_per_step,
+        DoEveryNEpisode() do t, agent, env
+            with_logger(lg) do
+                @info(
+                    "training",
+                    loss = agent.policy.loss,
+                    baseline_loss = agent.policy.baseline_loss,
+                    reward = total_reward_per_episode.rewards[end],
+                )
+            end
+        end,
+        DoEveryNEpisode(5000) do t, agent, env
+            RLCore.save(save_dir, agent)
+            BSON.@save joinpath(save_dir, "stats.bson") total_reward_per_episode time_per_step
+        end,
+    )
+
+    Experiment(
+        agent,
+        env,
+        StopAfterEpisode(5000),
+        hook,
+        Description("# Play Pendulum(Discrete) with VPG", save_dir),
+    )
+end
+
+function RLCore.Experiment(
+    ::Val{:JuliaRL},
+    ::Val{:VPG},
+    ::Val{:Pendulum},
+    ::Nothing;
+    save_dir = nothing,
+    seed = 123,
+)
+    # TODO: vpg does not work.
```

Comment:
How about removing this experiment for now given that it doesn't work yet?

Reply:
for the continuous action space experiment, I suspect there are bugs in the algo, or I used wrong hyper parameter.
Let me check it carefully.

Reply:
You are not alone. Haha https://discuss.pytorch.org/t/gym-pendulum-v0-not-solvable-by-vanilla-policy-gradient-increase-max-torques/8849/3

Let me know if you think it's ready to merge.

Reply:
VPG can solve discrete action space Pendulum in 5 sec, and the performance is stable.
![image](https://user-images.githubusercontent.com/2986988/93995620-d576bc80-fdc3-11ea-8283-d9d392af8d18.png)

But for continuous action space, VPG is very difficult to converge.
![image](https://user-images.githubusercontent.com/2986988/93995374-816bd800-fdc3-11ea-9faf-c6fb270137df.png)

the range of mean & std need to be selected carefully, or the agent can't learn anything.

and it needs a good seed.
I trained 12 agents.
6 agents can reach score -200.
3 agents can keep the average score > -300 in a few episodes.
but after that, the performance decreases gradually.
I guess it's the weakness of VPG, that it's variance is too large.

So I think there is no bug in the algo now.

Reply:
I just want to leave my small marks here:

1. Seeds should never be used to create good results. The reason this framework has a strong rng based implementation is to create reproducibility for single experiments (even though this is not guaranteed for major Julia updates). If the algorithm does only converge on a specific seed, then it probably needs some hyperparameter tuning somewhere else or is indeed unable to converge in this environment (as claimed [here](https://discuss.pytorch.org/t/gym-pendulum-v0-not-solvable-by-vanilla-policy-gradient-increase-max-torques/8849/3)) **The seed is no hyperparameter**

2. To be honest, Pendulum is also a bad environment to test and benchmark continuous control algorithms for 2 reasons: First, Pendulum performance is highly depended on the init values of the pendulum itself (>-100 performance seems only reachable if the pendulum spawns upright from the beginning). Second, pendulum only has a single dimension action space, which is often unrepresentative for real world control tasks, that's why you see Mujoco environments to benchmark the algorithms in most papers. I guess Pendulum is only used in this framework, because it is fairly easy to implement, fast and thereby very CI compatible.

I would therefore argue, that if you want to add this experiment, you should at least add a comment to the experiment which explains that this experimen only acts as a template implementation for the vpg in a continuous action space.

Reply:
Thanks for your advice.

I agree, choose a good seed is cheat.

But I need to make sure this code can work. Is there another continuous action space environment to use, which is very simple?

Reply:
Unfortunately, I have not come across a nice environment for this purpose. Guess, that's the reason why mujoco is so present in this field.

https://github.com/JuliaReinforcementLearning/ReinforcementLearningEnvironments.jl has an implementation of `ContinuousMountainCarEnv`. But that version is implemented with a very sparse reward function which I guess will make converging far more difficult. 

In theory there is also a continuous CartPole version, but I never managed to get an algorithm to converge in this env, not sure what's the issue there.

Finally, you can always fall back to some other Frameworks like OpenAI Gym (e.g. the Box2d environments), but these experiments should not be included in the tests of course due to the additional setup steps required.

Reply:
I tried the continuous version MountainCarEnv, I think it's impossible to converge. I logged the reward in every episode, it rarely got reward. so there's no gradient at all.

I also tried to use PG from https://github.com/thu-ml/tianshou, with Pendulum, it does not converge too.

In fact tianshou did not implement continuous action space case at all. I modified a few code, in order to use Gaussian policy agent.
PARL (https://github.com/PaddlePaddle/PARL/blob/develop/parl/algorithms/torch/policy_gradient.py) only supports Categorical distribution too.

So I guess VPG with a Gaussian policy is not practical.